libvcx/src/utils/devsetup_agent.rs

```diff
@@ -0,0 +1,462 @@
+
+pub const SERIALIZE_VERSION: &'static str = "2.0";
+
+#[cfg(test)]
+pub mod test {
+    use agency_client::payload::PayloadKinds;
+
+    use crate::{aries, connection, credential, credential_def, disclosed_proof, issuer_credential, proof, schema, settings};
+    use crate::aries::messages::a2a::A2AMessage;
+    use crate::error::{VcxError, VcxErrorKind, VcxResult};
+    use crate::libindy::utils::wallet::*;
+    use crate::utils::devsetup::*;
+    use crate::utils::provision::{provision_cloud_agent, ProvisionAgentConfig, AgencyConfig};
+    use crate::init::{open_as_main_wallet, init_issuer_config, create_agency_client_for_main_wallet};
+    use crate::utils::constants;
+
+    #[derive(Debug)]
+    pub struct VcxAgencyMessage {
+        pub uid: String,
+        pub decrypted_msg: String,
+    }
+
+    fn determine_message_type(a2a_message: A2AMessage) -> PayloadKinds {
+        debug!("determine_message_type >>> a2a_message={:?}", a2a_message);
+        match a2a_message.clone() {
+            A2AMessage::PresentationRequest(_) => PayloadKinds::ProofRequest,
+            A2AMessage::CredentialOffer(_) => PayloadKinds::CredOffer,
+            A2AMessage::Credential(_) => PayloadKinds::Cred,
+            A2AMessage::Presentation(_) => PayloadKinds::Proof,
+            _msg => PayloadKinds::Other(String::from("aries"))
+        }
+    }
+
+    fn str_message_to_a2a_message(message: &str) -> VcxResult<A2AMessage> {
+        Ok(serde_json::from_str(message)
+            .map_err(|err| VcxError::from_msg(VcxErrorKind::InvalidJson, format!("Cannot deserialize A2A message: {}", err)))?
+        )
+    }
+
+    fn str_message_to_payload_type(message: &str) -> VcxResult<PayloadKinds> {
+        let a2a_message = str_message_to_a2a_message(message)?;
+        Ok(determine_message_type(a2a_message))
+    }
+
+    fn download_message(did: String, filter_msg_type: PayloadKinds) -> VcxAgencyMessage {
+        let mut messages = agency_client::get_message::download_messages_noauth(Some(vec![did]), Some(vec![String::from("MS-103")]), None).unwrap();
+        assert_eq!(1, messages.len());
+        let messages = messages.pop().unwrap();
+
+        for message in messages.msgs.into_iter() {
+            let decrypted_msg = &message.decrypted_msg.unwrap();
+            let msg_type = str_message_to_payload_type(decrypted_msg).unwrap();
+            if filter_msg_type == msg_type {
+                return VcxAgencyMessage {
+                    uid: message.uid,
+                    decrypted_msg: decrypted_msg.clone(),
+                };
+            }
+        }
+        panic!("Message not found")
```

Comment:
Hmm, I know it is equivalent to returning a result and unwrapping, but this panic kind of breaches the contract I am tempted to have with a function which doesn't return a result

Reply:
Good point, moved panicking to upper layers of tests, instead returning `Option` here